### DISCUSSION TITLE
Clarification on user-defined constraint options

### DISCUSSION THREAD

**Discussion by lherrerai** [calliope-project/calliope#659] [Q&A] [answered]
Hello!

I'm working on a model (v0.6.10) where I'm trying to add a conversion technology that is tied to a storage technology, and would like to impose constraints on storage variables (such as charge, discharge, and energy level) for each timestep. Specifically, I want to set hourly limits on these variables throughout the simulation period.

I noticed the example on [user-defined custom constraints](https://calliope.readthedocs.io/en/stable/user/advanced_constraints.html#user-defined-custom-constraints) in the documentation, where the `max_capacity_90_constraint` is implemented using `energy_cap` and `energy_cap_max`. I want to ensure that I fully understand the range of options available.

Could you provide a list or explanation of the different parameters (such as `energy_cap`) that can be used in Calliope, especially for setting such constraints? Any insights or guidance would be greatly appreciated.

Thank you very much for your assistance!

**Comment by sjpfenninger** [comment, marked as answer]
In principle any of defined parameters and any variables, which will make their way into the Pyomo-built model, are available (see https://calliope.readthedocs.io/en/v0.6.10/user/config_defaults.html#per-tech-constraints and https://calliope.readthedocs.io/en/v0.6.10/user/ref_formulation.html#module-calliope.backend.pyomo.variables). 

I would recommend however moving to v0.7 instead, which has vastly superior ways to customise constraints:

https://calliope.readthedocs.io/en/latest/user_defined_math/
